Enable interconnect for localnet topology

jcaamano opened the pull request:
On L2 topologies such as localnet, there is no host subnet partitioning so each zone cannot independently allocate IPs to its local pods. While it would have been possible to have such partitioning, it would mean less than optimum IP allocation which is particularly important for localnet topology as it is a typical scenario to use public IP address spacing.

The main feature of this PR is a pod ip allocator for cluster manager. It does not only handle pod IP allocation, but fully handles the pod annotation as well (that also includes the MAC address, gateway and routes ) which then triggers the CNI side to attach the pod while the zone network controllers configure the NBDB. It is only enabled for localnet topology with interconnect and in such scenario the local zone network controller is inhibited from handling the pod annotation.

There is not much more required to enable interconnect for localnet topology. Specifically, this topology does not require a transit switch to interconnect the zones as it relies on the physical network to provide the connectivity between them.

Some refactoring happened for better code re-usability between cluster manager and the network controllers, being the most important:

- A subnet ip allocator was subtracted from the logical switch manager and made available for wider use so that IP allocation is not necessarily associated with the switch or node concepts. Logical switch manager is left around to perform the default allocation for gateway, management and hybrid overlay addresses that it was currently doing.
- The pod annotation handling code was also extracted from the `BaseNetworkController` into independent utility functions
- Most of cluster manager's network controller code which was handling host subnet allocation for the other topologies was moved to the host subnet allocator itself. This controller now also coordinates pod events.

**caveat 1**: There is potential for a race condition where an recently de-allocated IP address is allocated for a different pod while the zones have not had the chance to notice the de-allocation yet. This is mitigated by the round robin strategy of the allocator but could still happen when much of the address space is used up or when many allocations/de-allocations happen in a short period of time. Note that this is not specific to this PR and might happen as well with the default network on interconnect. The ultimate solution to this is to coordinate the de-allocation among all zones.

**caveat 2**: No interface hot plug support just as with the non-interconnect scenario.

~~**caveat 3**: enabling e2e test for localnet interconnect requires to bring up a suitable physcial network for it, which @maiqueb is currently working on~~

maiqueb commented:
@jcaamano maybe we should get https://github.com/ovn-org/ovn-kubernetes/pull/3621 in first so we can know for sure this feature works.

AFAIU this would remove caveat 3 on your list.

jcaamano commented:
> @jcaamano maybe we should get #3621 in first so we can know for sure this feature works.
> 
> AFAIU this would remove caveat 3 on your list.

I will try your changes locally but the order wouldn't matter because it is not going to run until PR https://github.com/ovn-org/ovn-kubernetes/pull/3366 is merged after

maiqueb commented:
> > @jcaamano maybe we should get #3621 in first so we can know for sure this feature works.
> > AFAIU this would remove caveat 3 on your list.
> 
> I will try your changes locally but the order wouldn't matter because it is not going to run until PR #3366 is merged after

I don't follow. My PR is working locally. 0_o

jcaamano commented:
> > > @jcaamano maybe we should get #3621 in first so we can know for sure this feature works.
> > > AFAIU this would remove caveat 3 on your list.
> > 
> > 
> > I will try your changes locally but the order wouldn't matter because it is not going to run until PR #3366 is merged after
> 
> I don't follow. My PR is working locally. 0_o

not with interconnect enabled?

maiqueb commented:
> > > > @jcaamano maybe we should get #3621 in first so we can know for sure this feature works.
> > > > AFAIU this would remove caveat 3 on your list.
> > > 
> > > 
> > > I will try your changes locally but the order wouldn't matter because it is not going to run until PR #3366 is merged after
> > 
> > 
> > I don't follow. My PR is working locally. 0_o
> 
> not with interconnect enabled?

oh of course! My bad. Thanks for explaining.

jcaamano commented:
/retest

jcaamano commented:
/retest

jcaamano commented:
/retest

maiqueb reviewed (commented):
Thank you for providing this functionality, and for refactoring this part of the code.

It is now more readable, and is also covered by unit tests, which is a very nice improvement.

I have some opinionated suggestions you can ignore, and some questions.

The PR was a bit hard to review, but now that I got the picture, next iterations should be simpler.

jcaamano commented on go-controller/pkg/ipallocator/allocator.go:
```diff
@@ -199,6 +200,17 @@ func (r *Range) Has(ip net.IP) bool {
 
+// Reserved returns true if the provided IP is reserved from allocation. This is
+// true for the network and broadcast addresses.
+func (r *Range) Reserved(ip net.IP) bool {
```
`Has` is to check if an IP has been previously allocated. `Reserved` is to check if an IP can be allocated or not. I will add the clarification you suggested.

jcaamano commented on go-controller/pkg/allocator/ip/subnet/allocator.go:
```diff
@@ -0,0 +1,292 @@
+}
+
+// AllocateUntilFull used for unit testing only, allocates the rest of the subnet
+func (allocator *SubnetIPAllocator) AllocateUntilFull(name string) error {
```
Can't because the function is used in tests out of package.

jcaamano commented on go-controller/pkg/util/pod_annotation.go:
```diff
@@ -371,3 +376,330 @@ func GetK8sPodAllNetworkSelections(pod *v1.Pod) ([]*nadapi.NetworkSelectionEleme
+	}
+	defer func() {
+		if err != nil {
+			rollback()
```
We return a rollback function to allow the possibility to rollback the allocation to this function user even when there was no error within. It is used in `AllocateToPodWithRollbackFunc` and explained in the function doc. Let me know how this could be clarified further.

maiqueb replied:
My question was misleading, sorry for that.

I meant to ask if we shouldn't return `nil` as rollback function when there **is** an error. All in all, you've already rolled it back. I guess it is up to the caller to never roll it back when it sees an error then.

You're good, as you said, your docstring explains quite well what this is supposed to do.

jcaamano replied:
Yeah, we could do that. But we do return something that can be used but does nothing so that callers are not forced to check for nil, which they should but I don't see that as ingrained in golang as let's say C. I thought there's no harm in returning a noop rollback.

maiqueb replied:
totally fine. 

Maybe the docstring should say the rollback functions must be idempotent ?

jcaamano commented on go-controller/pkg/util/pod_annotation.go:
```diff
@@ -371,3 +376,330 @@ func GetK8sPodAllNetworkSelections(pod *v1.Pod) ([]*nadapi.NetworkSelectionEleme
+		// If present, then we need to add default route for it
+		podAnnotation.Gateways = append(podAnnotation.Gateways, network.GatewayRequest...)
+		topoType := netinfo.TopologyType()
+		switch topoType {
+		case types.Layer2Topology, types.LocalnetTopology:
+			// no route needed for directly connected subnets
+			return nil
+		case types.Layer3Topology:
```
I see no difference other than it might provide more clarity so maybe but I just moved the code over as it was.

jcaamano commented on go-controller/pkg/util/pod_annotation_unit_test.go:
```diff
@@ -310,3 +317,462 @@ func TestGetPodIPsOfNetwork(t *testing.T) {
 	}
 }
+
+//lint:ignore U1000 stub
```
It's to ignore the `unused` linter warning but it was a temporary thing that is not actually needed so will remove

jcaamano commented on go-controller/pkg/util/pod_annotation_unit_test.go:
```diff
@@ -310,3 +317,462 @@ func TestGetPodIPsOfNetwork(t *testing.T) {
+		wantUpdatedPodAnnotation  *PodAnnotation
+		wantReleasedIPs           []*net.IPNet
+		wantReleasedIPsOnRollback []*net.IPNet
+		wantErr                   bool
```
I don't think it is that useful in this case:

- The function under test does not introduce any well-known error
- Checking for specific error strings is not really that helpful if nobody is actually depending on them.

jcaamano commented on go-controller/pkg/util/pod_annotation_unit_test.go:
```diff
@@ -310,3 +317,492 @@ func TestGetPodIPsOfNetwork(t *testing.T) {
+		name                      string
+		args                      args
+		ipam                      bool
+		invalidNetworkAnnotation  bool
```
I am trying to force an error while un-marshaling the network annotation. I can't do that with the network selection element in args.

jcaamano commented on go-controller/pkg/util/pod_annotation_unit_test.go:
```diff
@@ -310,3 +317,492 @@ func TestGetPodIPsOfNetwork(t *testing.T) {
+	tests := []struct {
+		name                      string
+		args                      args
+		ipam                      bool
```
Not really if I am just really interested in ipam vs no ipam and I don't care about any of the other details of the netInfo.

jcaamano commented on go-controller/pkg/util/pod_annotation_unit_test.go:
```diff
@@ -310,3 +317,492 @@ func TestGetPodIPsOfNetwork(t *testing.T) {
+	type args struct {
+		ipAllocator   ipAllocator
+		podAnnotation *PodAnnotation
+		network       *nadapi.NetworkSelectionElement
```
`AllocatePodAnnotationWithRollback` which is what is being tested here only takes one as argument.

jcaamano commented on go-controller/pkg/util/pod_annotation_unit_test.go:
```diff
@@ -310,3 +317,492 @@ func TestGetPodIPsOfNetwork(t *testing.T) {
+		args                      args
+		ipam                      bool
+		invalidNetworkAnnotation  bool
+		wantGeneratedMac          bool
+		wantUpdatedPodAnnotation  *PodAnnotation
+		wantReleasedIPs           []*net.IPNet
+		wantReleasedIPsOnRollback []*net.IPNet
```
this is the terminology that the gotests stub generator uses so I think it is better to not be nit picky about this or we will force everyone making use of it to change stuff around.

jcaamano commented on go-controller/pkg/util/pod_annotation_unit_test.go:
```diff
@@ -310,3 +317,492 @@ func TestGetPodIPsOfNetwork(t *testing.T) {
+			if network == nil {
+				network = &nadapi.NetworkSelectionElement{}
+			}
+			network.Name = "network"
+			network.Namespace = "namespace"
```
Because it is the same for all the tests.

Can you elaborate where do you suggest to call that from?

jcaamano replied:
Not sure about this anyway, I see this it very specific for this test case.

jcaamano commented on go-controller/pkg/util/pod_annotation_unit_test.go:
```diff
@@ -310,3 +317,492 @@ func TestGetPodIPsOfNetwork(t *testing.T) {
+			network.Name = "network"
+			network.Namespace = "namespace"
+
+			pod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "pod",
+					Namespace: "namespace",
+				},
+			}
+
+			if tt.invalidNetworkAnnotation {
+				pod.ObjectMeta.Annotations = map[string]string{
+					nadapi.NetworkAttachmentAnnot: "",
+				}
+			}
```
Not sure about this, I see this it very specific for this test case.

maiqueb replied:
as above, fair enough.

jcaamano commented on go-controller/pkg/util/pod_annotation.go:
```diff
@@ -371,3 +380,326 @@ func GetK8sPodAllNetworkSelections(pod *v1.Pod) ([]*nadapi.NetworkSelectionEleme
+// IP allocation if there was one. This function is designed to be used in
+// AllocateToPodWithRollbackFunc implementations. Use an inlined implementation
+// if you want to extract information from it as a side-effect.
+func AllocatePodAnnotationWithRollback(
```
Don't like `GeneratePodAnnotation` either.

A PodAnnotation is allocated by extension of an IP being allocated and all the contained information being derived from that allocation in a way that two running pods would not have the same PodAnnotation.

The meaning of Allocation here comes with the added connotation that you can `undo` through the rollback and I think it is important to carry that connotation forward without trying to read to much into it otherwise.

It might not be perfect and  I will take further suggestions to try to come up with a better name but I think this is or could be sufficiently explained in the doc of the method.

jcaamano commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -42,80 +33,66 @@ type networkClusterController struct {
-	ncc.retryNodes = ncc.newRetryFramework(factory.NodeType, true)
+func (ncc *networkClusterController) hasNodeSubnetAllocation() bool {
+	// we only do node subnet allocation on L3 topologies or default network
+	return ncc.TopologyType() == types.Layer3Topology || !ncc.IsSecondary()
```
The topology of the primary network is not `types.Layer3Topology`

maiqueb replied:
hm, how about that.. I thought it was. 

It doesn't have a topology set then ?

jcaamano commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -527,8 +290,7 @@ func (h *networkClusterControllerEventHandler) GetResourceFromInformerCache(key
 
 	switch h.objType {
-	case factory.NodeType,
-		factory.EgressNodeType:
```
I guess this was here by mistake. No retry framework is being setup for `factory.EgressNodeType` in network cluster controller so we won't ever get an event for this type.

maiqueb replied:
thanks for explaining, and for correcting.

tssurya replied:
yea this is a leftover somehow from when the main IC changes merged; : https://github.com/ovn-org/ovn-kubernetes/pull/3127#discussion_r1089986949

jcaamano commented on go-controller/pkg/clustermanager/subnetallocator/host_subnet_allocator.go:
```diff
@@ -3,57 +3,329 @@ package subnetallocator
-	v4count, _, v6count, _ := sna.base.Usage()
-	metrics.RecordSubnetCount(float64(v4count), float64(v6count))
+	if sna.hasHybridOverlayAllocation() {
+		for _, hoSubnet := range config.HybridOverlay.ClusterSubnets {
```
it's hybrid overlay subnet, not host subnet

maiqueb replied:
darn. Is this commonly used elsewhere ? If not, it could be a bit confusing. At least it confused me.

jcaamano commented on go-controller/pkg/clustermanager/subnetallocator/host_subnet_allocator.go:
```diff
@@ -3,57 +3,329 @@ package subnetallocator
+	}
+
+	err := sna.syncNodeNetworkAnnotations(node)
+	sna.recordSubnetUsage()
```
I don't think this is an issue since it pulls from a consistent state of the allocator

jcaamano commented on go-controller/pkg/clustermanager/subnetallocator/host_subnet_allocator.go:
```diff
@@ -3,57 +3,329 @@ package subnetallocator
+	// were already assigned to the node. allocatedSubnets will contain
+	// any newly allocated subnets required to ensure that the node has one subnet
+	// from each enabled IP family.
+	ipv4Mode, ipv6Mode := sna.netInfo.IPMode()
+	validExistingSubnets, allocatedSubnets, err := sna.allocateNodeSubnets(sna.clusterSubnetAllocator, node.Name, existingSubnets, ipv4Mode, ipv6Mode)
```
We don't use the info from sna.netInfo for hybrid overlay allocations, those are always ipv4 only. I guess if hybrid overlay is enabled, sna.netInfo should also have ipv6 disabled but it is being hardcoded for extra carefulness.

maiqueb replied:
I don't follow, sorry. 

I'm just asking if you can't implement 

```golang
func (sna *HostSubnetAllocator) allocateNodeSubnets(allocator SubnetAllocator, nodeName string, existingSubnets []*net.IPNet) ([]*net.IPNet, []*net.IPNet, error) {
    ipv4Mode, ipv6Mode := sna.netInfo.IPMode()
...
```

instead of 
```golang
func (sna *HostSubnetAllocator) allocateNodeSubnets(allocator SubnetAllocator, nodeName string, existingSubnets []*net.IPNet, ipv4Mode, ipv6Mode bool) ([]*net.IPNet, []*net.IPNet, error) {
```

Anyway, this is just a nit intending to avoid passing two extra arguments to the function.

jcaamano replied:
No because in line 113 is being called as
```
hostSubnets, allocatedSubnets, err := sna.allocateNodeSubnets(sna.hybridOverlaySubnetAllocator, node.Name, existingSubnets, true, false)
```
because of the reason I explained above.

maiqueb replied:
got it.

jcaamano commented on go-controller/pkg/clustermanager/subnetallocator/host_subnet_allocator_test.go:
```diff
@@ -258,153 +277,49 @@ func TestController_allocateNodeSubnets_ReleaseOnError(t *testing.T) {
-	return slice, nil
-}
-
-func TestController_markSubnetsAllocated(t *testing.T) {
```
Not needed since I removed the `markSubnetsAllocated` which was just a passthrough to SubnetAllocator `MarkAllocatedNetworks` which is already being tested by itself

maiqueb replied:
tks for explaining.

jcaamano commented:
> Thank you for providing this functionality, and for refactoring this part of the code.
> 
> It is now more readable, and is also covered by unit tests, which is a very nice improvement.
> 
> I have some opinionated suggestions you can ignore, and some questions.
> 
> The PR was a bit hard to review, but now that I got the picture, next iterations should be simpler.

Thanks. Please note that a lot of comments apply for pieces of code that I had just moved over to a different place. I will follow your suggestions where it makes since since you made the effort to take a look at it.

jcaamano commented:
Added a new commit at the end. See commit msg.

@maiqueb FYI

jcaamano commented:
> ~~Added a new commit at the end. See commit msg.~~
> 
> 
> ~~@maiqueb FYI~~
> 
> Scratch that, I need to drop the last commit, it's not going to work as I would hope for

Dropped for now. Nothing that is strictly needed for this PR.

maiqueb commented:
/retest-failed

jcaamano commented:
Added a new commit. This is not the same as the previous commit I dropped. I could have squashed it but want to give a heads up to @maiqueb first as he already reviewed.

maiqueb commented:
> Added a new commit. This is not the same as the previous commit I dropped. I could have squashed it but want to give a heads up to @maiqueb first as he already reviewed.

Thank you ! I'll provide feedback today.

maiqueb commented:
> > Added a new commit. This is not the same as the previous commit I dropped. I could have squashed it but want to give a heads up to @maiqueb first as he already reviewed.
> 
> Thank you ! I'll provide feedback today.

I think you can squash it @jcaamano , this is better now.

maiqueb commented on go-controller/pkg/clustermanager/pod/ip_allocator.go:
```diff
@@ -34,20 +34,16 @@ type PodIPAllocator struct {
 	kube      kube.Interface
-
-	// function used to update the pod, replaceable mainly for testing purposes
-	updatePodWithAllocationOrRollback func(listers.PodLister, kube.Interface, *corev1.Pod, util.AllocateToPodWithRollbackFunc) error
```
could you explain why we're not doing the same for DPU ? 
@jcaamano

jcaamano replied:
hmm not sure I understand in the context of the code you are highlighting. The same as what?

maiqueb replied:
I mean that DPU still uses the code you said you were removing in the top-most commit: 
https://github.com/ovn-org/ovn-kubernetes/pull/3620/files#diff-2e2e03f067b90c926504d0f11b39ed85685ffc2f0f98de26d69bc08c490e9bddR222

Quoting your top-most commit msg:
> Unexport AllocatePodAnnotationWithRollback and add AllocatePodAnnotation
which takes no PodAnnotation argument and only returns a PodAnnotation
the pod has already been updated with.

Thus I assumed "everyone" would use this new form.

maiqueb replied:
After writing my msg above I noticed that `UpdatePodAllocationWithRetryOrRollback` != `AllocatePodAnnotationWithRollback` .

jcaamano replied:
Its `allocatePodAnnotationWithRollback` now, not exported so nobody of the util package can use that without some kungfu

jcaamano replied:
Changed `UpdatePodAllocationWithRetryOrRollback` to `UpdatePodWithRetryOrRollback`. No point in being so precise with the naming if it brings confusion. Let me know what you think.

jcaamano replied:
github is not picking up the latest push for some reason. We might just need to be patient.

maiqueb replied:
thank you. I do prefer that name.

jcaamano commented:
/retest

jcaamano commented:
Added some fixes for multi-homing e2e tests. The underlay was not being properly configured. The only explanation I have for the localnet tests working in non-IC configuration is that as long as there is no binding for the localnet port, the switch behaves as a normally distributed switch.

@maiqueb PTAL

maiqueb commented on go-controller/pkg/clustermanager/secondary_network_cluster_manager.go:
```diff
@@ -94,7 +94,8 @@ func (sncm *secondaryNetworkClusterManager) Stop() {
 func (sncm *secondaryNetworkClusterManager) NewNetworkController(nInfo util.NetInfo) (nad.NetworkController, error) {
 	topoType := nInfo.TopologyType()
-	if topoType == ovntypes.Layer3Topology {
+	klog.Infof("Creating new network constroller for network %s of topology %s", nInfo.GetNetworkName(), topoType)
```
```suggestion
	klog.Infof("Creating new network controller for network %s of topology %s", nInfo.GetNetworkName(), topoType)
```

maiqueb commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -172,6 +224,20 @@ func (h *networkClusterControllerEventHandler) UpdateResource(oldObj, newObj int
+		if !ok {
+			return fmt.Errorf("could not cast %T new object to *corev1.Pod", newObj)
+		}
+		err := h.ncc.podIPAllocator.Reconcile(new, old)
```
aren't the old / new pod arguments switched ? The function's signature is:
```golang
func (a *PodIPAllocator) Reconcile(old, new *corev1.Pod) error {
...
```
Src: https://github.com/ovn-org/ovn-kubernetes/pull/3620/files#diff-d0111b0bbab46d1b09ce2dcb9bf69046b81311a60da03baad0fce16c0c693d2dR74

jcaamano replied:
:facepalm:

maiqueb commented on go-controller/pkg/clustermanager/pod/ip_allocator.go:
```diff
@@ -0,0 +1,265 @@
+
+	if podDeleted {
+		a.deleteReleasedPod(nad, string(pod.UID))
+	} else {
+		a.addReleasedPod(nad, string(pod.UID))
+	}
+
+	return nil
```
nit

```suggestion
	} 
	return a.addReleasedPod(nad, string(pod.UID))
```

jcaamano replied:
Can't do. `addReleasedPod`  does not return anything.

maiqueb replied:
ah! my turn to facepalm :)

maiqueb commented on go-controller/pkg/clustermanager/pod/ip_allocator_test.go:
```diff
@@ -0,0 +1,342 @@
+	if p.scheduled {
+		pod.Spec.NodeName = "node"
+	}
+	if p.hostNetwork {
+		pod.Spec.HostNetwork = true
+	}
```
nit: `pod.Spec.HostNetwork = p.hostNetwork`

maiqueb reviewed (commented):
Looking good.

Most of what I have are nits.

jcaamano commented on test/e2e/localnet-underlay.go:
```diff
@@ -156,6 +157,9 @@ func defaultNetworkBridgeMapping() BridgeMapping {
 }
 
 func bridgeMapping(physnet, ovsBridge string) BridgeMapping {
+	physnet = strings.ReplaceAll(physnet, "-", ".")
+	physnet = strings.ReplaceAll(physnet, "/", ".")
+	physnet = physnet + localnetSuffix
```
I would prefer not to do that. There is not a single import to production code from e2e.

jcaamano commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -60,10 +67,23 @@ func newNetworkClusterController(networkID int, netInfo util.NetInfo, ovnClient
 }
 
+func (ncc *networkClusterController) hasPodIPAllocation() bool {
+	// we only do pod IP allocation on L2 topologies with IPAM on interconnect
+	switch ncc.TopologyType() {
+	case types.Layer2Topology, types.LocalnetTopology:
```
That is already handled in the secondary network cluster manager. I prefer this bit to be in its final form.

maiqueb replied:
Fine with me.

maiqueb reviewed (approved):
Looks good.

After this the code is more modular, and covered by unit tests. Thank you.

maiqueb commented:
I confirm this last version works for podman (adding `--internal` for podman deployments).
@jcaamano

jcaamano commented:
/retest-failed

trozet commented on go-controller/pkg/ovn/base_network_controller_pods.go:
```diff
@@ -611,138 +491,15 @@ func (bnc *BaseNetworkController) addLogicalPortToNetwork(pod *kapi.Pod, nadName
-			return nil, nil, nil, false, err
-		}
-		releaseIPs = false
+	podAnnotation, annotationUpdated, err := bnc.allocateIPs(pod, existingLSP, nadName, network)
```
I feel like you are replacing more than just "allocateIPs" as that logic also configures routes for the pod, mac, gateway ips. Maybe it should be allocateNetworkingForPod or something like that?

trozet commented on go-controller/pkg/ovn/base_network_controller_pods.go:
```diff
@@ -611,138 +491,15 @@ func (bnc *BaseNetworkController) addLogicalPortToNetwork(pod *kapi.Pod, nadName
-		}
-		releaseIPs = false
+	podAnnotation, annotationUpdated, err := bnc.allocateIPs(pod, existingLSP, nadName, network)
+	if err != nil {
```
so these changes are very scary imo..it took us a long time to get the addLogicalPort path solid, and I'm pretty hesitant to make any change here. I would prefer you just change the path for localnet and leave primary network alone.

jcaamano replied:
OK, I can do that. 

But I have thoughts about this.

If we cannot refactor this code, then that makes the code unmaintainable from my perspective. I respected all the comments the code had and did not brake any tests. In exchange, I added more comments and tests. Yes, there might be regressions for things which we had no comments or no tests, but if we ever want to improve this, we need to take the plunge.

trozet replied:
I agree that we should be able to refactor it at some point, I just dont think the time is now with all of the other risks and code change related to IC. When we had issues with the addLogicalPort stuff in the past it was very complicated to debug and go back in time and figure out exactly what happened to cause the problems. I dont want to risk going down that path while trying to test out our new architecture in downstream. I would vote we have branches for now (one primary one everything else) and then after we get past stabilizing IC, we consolidate to the 1 path.

That will also give us some time to make sure the changes you make work well for secondary networks before we try it on primary.

jcaamano replied:
I moved the old allocation for the default network to `allocatePodAnnotation` and move the new allocation for secondary networks to `allocatePodAnnotationForSecondaryNetwork`.

Keeping the old functionality in the original addLogicalPortToNetwork would have just been messy, so I tried to reach you midway.

The old functionality was moved intact. The only differences are
-  I removed `lsp.DynamicAddresses = nil`, we were not writing that as `DynamicAddresses` was not in the list of updated fields we pass to libovsdb client.
- `AddRoutesGatewayIP` still lives in `util/pod_annotation.go` so that it can be reused. I carries a change I introduced to derive the subnets from the pod IPs but is otherwise intact.

trozet commented on go-controller/pkg/ovn/base_network_controller_pods.go:
```diff
@@ -920,30 +639,62 @@ func (bnc *BaseNetworkController) WatchPods() error {
+		ipAllocator = bnc.lsManager.ForSwitch(switchName)
+	}
+
+	updatedPod, podAnnotation, err := util.AllocatePodAnnotation(
```
seems like we are passing a lot of arguments to this function with bnc properties. Why not make it a receiver?

jcaamano replied:
I can make a new struct for it that we initialize with some of the args and can act as receiver. WDYT?

trozet replied:
I was thinking we can just make some new interface that bnc inherits and ovn_cluster_manager can also use. I guess  maybe it is similar to the interface you have in util/pod_annotation.go. Then we can make struct that implements that interface, and make these functions be receivers to that.

jcaamano replied:
That seem to it would be inheriting just to share code. There should be a clear composition relationship for that. We can just have an utility struct and use it where needed.

trozet commented on go-controller/pkg/util/pod_annotation.go:
```diff
@@ -379,3 +389,415 @@ func GetK8sPodAllNetworkSelections(pod *v1.Pod) ([]*nadapi.NetworkSelectionEleme
+
+// ipAllocator is an IP allocation interface that needs to be implemented to use
+// AllocatePodAnnotation
+type ipAllocator interface {
```
we already have allocator defined in allocator package, and now we are defining another one in the util package? It seems like we are building more than utility functions here, and I'm afraid its making things more complicated.

jcaamano replied:
This type is not exported and you should only care about it if you use the function that needs it. We don't have an equivalent interface exported anywhere else.

But I can certainly define one and export it from the ip allocator package itself.

trozet reviewed (commented):
reviewed through: Add PodAnnotation update utility functions

will continue reviewing tmrw...but in general I dont like how we are moving stuff to the util package that seems like it should be another type in another package that cluster-manager and default network controller can both create instances of. I also left comments about not changing the primary network code path as I really dont want to risk anything at all there.

jcaamano commented on go-controller/hack/test-go.sh:
```diff
@@ -59,7 +59,7 @@ function testrun {
         args="${args} -test.timeout=20m"
     fi
-    if grep -q -r "ginkgo" ."${path}"; then
+    if grep -q "ginkgo" ."${path}"/*_test.go; then
```
With the change:
```
gocmd test -mod=vendor -test.v -race  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/allocator 
```
Without the change
```
gocmd test -mod=vendor -test.v -race  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/allocator -ginkgo.v  -ginkgo.reportFile ./_artifacts/junit-pkg_allocator.xml
flag provided but not defined: -ginkgo.v
Usage of /tmp/go-build1574099407/b001/allocator.test:
  -test.bench regexp
        run only benchmarks matching regexp
  -test.benchmem
        print memory allocations for benchmarks
  -test.benchtime d
        run each benchmark for duration d (default 1s)
  -test.blockprofile file
        write a goroutine blocking profile to file
  -test.blockprofilerate rate
        set blocking profile rate (see runtime.SetBlockProfileRate) (default 1)
  -test.count n
        run tests and benchmarks n times (default 1)
  -test.coverprofile file
        write a coverage profile to file
  -test.cpu list
        comma-separated list of cpu counts to run each test with
  -test.cpuprofile file
        write a cpu profile to file
  -test.failfast
        do not start new tests after the first test failure
  -test.fuzz regexp
        run the fuzz test matching regexp
  -test.fuzzcachedir string
        directory where interesting fuzzing inputs are stored (for use only by cmd/go)
  -test.fuzzminimizetime value
        time to spend minimizing a value after finding a failing input (default 1m0s)
  -test.fuzztime value
        time to spend fuzzing; default is to run indefinitely
  -test.fuzzworker
        coordinate with the parent process to fuzz random values (for use only by cmd/go)
  -test.list regexp
        list tests, examples, and benchmarks matching regexp then exit
  -test.memprofile file
        write an allocation profile to file
  -test.memprofilerate rate
        set memory allocation profiling rate (see runtime.MemProfileRate)
  -test.mutexprofile string
        write a mutex contention profile to the named file after execution
  -test.mutexprofilefraction int
        if >= 0, calls runtime.SetMutexProfileFraction() (default 1)
  -test.outputdir dir
        write profiles to dir
  -test.paniconexit0
        panic on call to os.Exit(0)
  -test.parallel n
        run at most n tests in parallel (default 4)
  -test.run regexp
        run only tests and examples matching regexp
  -test.short
        run smaller test suite to save time
  -test.shuffle string
        randomize the execution order of tests and benchmarks (default "off")
  -test.testlogfile file
        write test action log to file (for use only by cmd/go)
  -test.timeout d
        panic test binary after duration d (default 0, timeout disabled)
  -test.trace file
        write an execution trace to file
  -test.v
        verbose: print additional output
FAIL    github.com/ovn-org/ovn-kubernetes/go-controller/pkg/allocator   0.013s
FAIL
```

trozet replied:
I mean I tested your change, and it did not what or how the tests were run. In other words, your change is good.

jcaamano replied:
Thanks for looking into that.

jcaamano commented on go-controller/pkg/allocator/ip/subnet/allocator.go:
```diff
@@ -0,0 +1,292 @@
+}
+
+// AddSubnet set to the allocator for IPAM management, or update it.
+func (allocator *SubnetIPAllocator) AddSubnet(name string, subnets []*net.IPNet) error {
```
Ok. Do you want to do the same to the logical switch manager `AddSwitch` to `AddOrUpdateSwitch`?

trozet replied:
yah

jcaamano replied:
Will add a new commit for this change

jcaamano replied:
No new commit at the end. Incorporated on their respective original commits.

jcaamano commented on go-controller/pkg/ovn/base_secondary_layer2_network_controller.go:
```diff
@@ -298,22 +297,9 @@ func (oc *BaseSecondaryLayer2NetworkController) InitializeLogicalSwitch(switchNa
 	}
 
-	if err = oc.lsManager.AddSwitch(switchName, hostSubnets); err != nil {
+	if err = oc.lsManager.AddSwitch(switchName, hostSubnets, excludeSubnets...); err != nil {
```
But that's all the commit is about and the commit message just talks about that:
```
Pre-allocate subnets in subnet ip allocator

Add the ability to pre-allocate subnets to the subnet ip allocator which
might be a common need.

Note that this is just a pre-allocation. The subnet ip allocator does
not distinguish between pre-allocated IPs vs normally allocated IPs and
will return the same ErrAllocated error on re-allocation.

In the process, refactor GetIPFullMask to be based on net.IPMask and
avoided parsing in a few places.
```

Perhaps I did not explain myself well but the whole intention of the commit and the commit message is this specific change.

trozet replied:
yeah you wrote "might be a common need", i think it would be clearer to give the context that you can provide excludedSubnets via NAD (or wherever we allow it) and it is common for localnet because the network is a shared external network where probably a bunch of the address space is already in use

jcaamano replied:
Reworded the commit message.

jcaamano commented:
> but in general I dont like how we are moving stuff to the util package that seems like it should be another type in another package that cluster-manager and default network controller can both create instances of. 

Got it. Will make a separate pkg and utility for it.

trozet commented on go-controller/pkg/clustermanager/pod/ip_allocator.go:
```diff
@@ -0,0 +1,265 @@
+
+	// track pods that have been released but not deleted yet so that we don't
+	// release more than once
+	releasedPods      map[string]sets.Set[string]
```
not sure why this is needed. Cant you just search the informer cache for any other pod that is not completed that is not using the IP and then release it if not? (similar to what primary path does today)

I guess maybe the cost will be too high, cause the secondary IP is stored in the annotation and not in the pod status ips? So we will have to json unmarshal a bunch of times?

trozet commented on go-controller/pkg/factory/factory.go:
```diff
@@ -535,6 +535,13 @@ func NewClusterManagerWatchFactory(ovnClientset *util.OVNClusterManagerClientset
 	}
 
+	if config.OVNKubernetesFeature.EnableInterconnect && config.OVNKubernetesFeature.EnableMultiNetwork {
+		wf.informers[PodType], err = newQueuedInformer(PodType, wf.iFactory.Core().V1().Pods().Informer(), wf.stopChan, defaultNumEventQueues)
```
not sure if we need a queued informer with 15  workers here, the amount of work being done is very small, and you have a giant lock on the cache you use anyway.

jcaamano replied:
The (new) lock just affects releases, not allocations, and just to lookup in the cache.

The work is indeed small, but incurs in the IO of updating the pod annotation (with retries).

Since you are not sure, and now I am not sure either, let's keep it as is and have a discussion about it and make any changes on a separate PR

trozet reviewed (changes requested):
Overall rest of it looks pretty good. Left some comments.

jcaamano commented on go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go:
```diff
@@ -1,109 +1,54 @@
-		if err != nil {
-			klog.Errorf("IPAM for subnet %s was not initialized for switch %q", subnet, switchName)
-			return err
+	if manager.reserveIPs {
```
I did not act on this at the end.

The list of IPs to reserve depends on the subnets for the switches dynamically added to the logical switch manager, so it does not make sense to make them a property of the logical switch manager itself.

I do think it might make sense to get rid of the logical switch manager alltogether  and handle IP reservation by passing the exclude subnets directly from the network controller to the subnet allocator. But this was a rework I did not want to do as part of this PR.

jcaamano commented:
> Looks like the path is:
NewClusterManager -> newEgressIPController(ovnClient, wf, recorder) -> initRetryFramework() -> eIPC.newRetryFramework(factory.EgressNodeType)

> So I guess Cluster Manager uses it, but Network Cluster Controller does not, correct?

Correct. From my point of view it was even m,ore clear since the egressip controller was not there yet when I did this work.

jcaamano commented on go-controller/pkg/clustermanager/subnetallocator/host_subnet_allocator.go:
```diff
@@ -3,57 +3,329 @@ package subnetallocator
+}
+
+// updateNodeNetworkAnnotationsWithRetry will update the node's subnet annotation and network id annotation
+func (sna *HostSubnetAllocator) updateNodeNetworkAnnotationsWithRetry(nodeName string, hostSubnetsMap map[string][]*net.IPNet, networkId int) error {
```
Yeah, that's what I did.

jcaamano commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -247,8 +325,9 @@ func (h *networkClusterControllerEventHandler) IsResourceScheduled(obj interface
-// IsObjectInTerminalState returns true if the object is a in terminal state.  Always returns true.
+// IsObjectInTerminalState returns true if the object is a in terminal state.  Always returns false.
 func (h *networkClusterControllerEventHandler) IsObjectInTerminalState(obj interface{}) bool {
+	// Note: the pod IP allocator needs to be aware when pods are deleted
```
Precisely. That is why we want to consider all pod states which is what omitting this implementation achieves.

The retry framework uses this to replace an update with a delete for objects in terminal state and inhibit deletes from being notified downstream in that case afterwards. We don't want that for the allocator. The allocator needs to be distinctly aware of when a pod becomes terminal vs when it is deleted; and it does keep track of that.

jcaamano commented:
> > but in general I dont like how we are moving stuff to the util package that seems like it should be another type in another package that cluster-manager and default network controller can both create instances of.
> 
> Got it. Will make a separate pkg and utility for it.

OK so this rework required some elaboration. Relevant changes were, by commit:

- *Re-structure ipallocator package*: 
  - additionally moved the bitmap allocator to its own package `pkg/allocator/bitmap`
- *Add PodAnnotation update utility*: 
  - moved the new code to build & set the PodAnnotation from `pkg/util/pod_annotation.go` to its own utility in `pkg/allocator/pod/pod_annotation.go`
  - moved the utility functions to generically update a pod with retry from `pkg/util/pod_annotation.go` to `pkg/util/pod.go`
  - exported a new `subnet.NamedAllocator` interface instead of the unexported one defined previously in `pkg/util/pod_annotation.go`
- *Add Named ID allocator*:
  -  exported a new `id.NamedAllocator` interface instead of the unexported one defined previously in `pkg/util/pod_annotation.go`
